src/main/java/org/openstreetmap/atlas/geography/atlas/raw/creation/OsmPbfCounter.java

```diff
@@ -385,17 +398,43 @@ private boolean shouldLoadOsmWay(final Entity entity)
         return this.loadingOption.isLoadOsmWay() && entity instanceof Way;
     }
 
-    private boolean wayContainsNodeWithinBoundary(final Way way)
+    private boolean wayIntersectsBoundary(final Way way)
     {
+        // CASE 1: Line crosses (or is enclosed by) the shard bounds and has at least one shapepoint
+        // within the shard bounds
+        final ArrayList<Location> wayNodesLocations = new ArrayList<>();
         for (final WayNode node : way.getWayNodes())
         {
+            // nodes are processed first so allNodes will contain all node locations
+            wayNodesLocations.add(this.nodeIdToLocation.get(node.getNodeId()));
             if (this.nodeIdentifiersToInclude.contains(node.getNodeId()))
             {
                 this.wayIdentifiersToInclude.add(way.getId());
                 return true;
             }
         }
 
+        // CASE 2: Line crossed the shard but has no shapepoints within it, so we must check for
+        // intersections
+        wayNodesLocations.stream().filter(node -> node != null);
+        if (wayNodesLocations.isEmpty())
+        {
+            return false;
+        }
+        final PolyLine wayGeometry = new PolyLine(wayNodesLocations);
+        if (wayGeometry.isPoint() || wayGeometry.isEmpty() || wayGeometry.bounds() == null)
+        {
+            return false;
+        }
+        // Checking the bounds of the polyline instead of the actual geometry may include some
```

Comment:
this is a great way of handling the performance requirements here! 👍

Reply:
Thanks, it was @matthieun's idea!